src/shell/shell.cpp

```diff
@@ -178,141 +178,269 @@ DOS_Shell::DOS_Shell()
+			// Overwrite with >, and append with >>
+			if (ch == '>' && (*append = (*lr == '>')))
+				lr++;
+			redir = ch == '>' ? ofn : (ch == '<' ? ifn : pipe);
+			if (*redir) {
+				delete[] * redir;
+				*redir = nullptr;
 			}
-			*ofn = lr;
-			while (*lr && *lr!=' ' && *lr!='<' && *lr!='|') lr++;
-			//if it ends on a : => remove it.
-			if((*ofn != lr) && (lr[-1] == ':')) lr[-1] = 0;
-			temp_len = static_cast<size_t>(lr - *ofn + 1u);
-			temp = new char[temp_len];
-			safe_strncpy(temp, *ofn, temp_len);
-			*ofn = temp;
+			lr = ltrim(lr);
+			str = *redir = lr;
+			chrs = ch == '|' ? "" : (ch != '<' ? " |<" : " |>");
+			found = str.find_first_of(chrs);
+			if (found == std::string::npos)
+				temp_len = str.size();
+			else
+				temp_len = found - (str[found - 1] == ':' ? 1 : 0);
+			lr += temp_len;
+			str[temp_len] = 0;
+			temp = new char[temp_len + 1];
+			safe_strncpy(temp, str.c_str(), temp_len + 1);
+			*redir = temp;
 			continue;
+		}
+		*lw++ = ch;
+	}
+	*lw = 0;
+}
 
-		case '<':
-			if (*ifn) {
-				delete[] * ifn;
-				*ifn = nullptr;
+bool get_pipe_status(const char *out,
+                     const char *pipe,
+                     char *pipetmp,
+                     const bool append,
+                     bool &failed_pipe)
+{
+	uint16_t fattr = 0;
+	uint16_t dummy = 0;
+	uint16_t dummy2 = 0;
+	uint32_t bigdummy = 0;
+	bool status = true;
+	/* Create if not exist. Open if exist. Both in read/write mode */
+	if (!pipe && append) {
+		if (DOS_GetFileAttr(out, &fattr) && fattr & DOS_ATTR_READ_ONLY) {
+			DOS_SetError(DOSERR_ACCESS_DENIED);
+			status = false;
+		} else if ((status = DOS_OpenFile(out, OPEN_READWRITE, &dummy))) {
+			DOS_SeekFile(1, &bigdummy, DOS_SEEK_END);
+		} else {
+			// Create if not exists.
+			status = DOS_CreateFile(out, DOS_ATTR_ARCHIVE, &dummy);
+		}
+	} else if (!pipe && DOS_GetFileAttr(out, &fattr) &&
+	           (fattr & DOS_ATTR_READ_ONLY)) {
+		DOS_SetError(DOSERR_ACCESS_DENIED);
+		status = false;
+	} else {
+		if (pipe && DOS_FindFirst(pipetmp, ~DOS_ATTR_VOLUME) &&
+		    !DOS_UnlinkFile(pipetmp))
+			failed_pipe = true;
+		status = DOS_OpenFileExtended(pipe && !failed_pipe ? pipetmp : out,
+		                              OPEN_READWRITE, DOS_ATTR_ARCHIVE,
+		                              0x12, &dummy, &dummy2);
+		if (pipe && (failed_pipe || !status) &&
+		    (Drives[0] || Drives[2]) && !strchr(pipetmp, '\\')) {
+			int len = (int)strlen(pipetmp);
+			const int pipetmp_limit = 266;
+			if (len > pipetmp_limit) {
+				len = pipetmp_limit;
+				pipetmp[len] = 0;
 			}
-			lr = ltrim(lr);
-			*ifn = lr;
+			for (int i = len; i >= 0; i--)
+				pipetmp[i + 3] = pipetmp[i];
+			pipetmp[0] = Drives[2] ? 'c' : 'a';
+			pipetmp[1] = ':';
+			pipetmp[2] = '\\';
+			failed_pipe = false;
+			if (DOS_FindFirst(pipetmp, ~DOS_ATTR_VOLUME) &&
+			    !DOS_UnlinkFile(pipetmp))
+				failed_pipe = true;
+			else
+				status = DOS_OpenFileExtended(pipetmp, OPEN_READWRITE,
```

Comment:
There's `pipe` and `pipetmp`  arguments into the function. 
I'm actually seeing more work done w/ pipetmp than pipe? 
I can't tell what they are, and why they're different - and why is one a tmp and other the 'real' one.

Reply:
I renamed the said variables to be easier to understand, and added comment to mention that piping requires the use of a temporary file, which is why the temp file variable is used for piping.